library/StandardLibrary/src/Juvix/Library/Feedback.hs

```diff
@@ -0,0 +1,101 @@
+module Juvix.Library.Feedback
+  ( Feedback (..),
+    FeedbackT (..),
+    feedback,
+    feedbackT,
+    Fail.fail,
+  )
+where
+
+import qualified Control.Monad.Fail as Fail
+import qualified Control.Monad.Trans as Trans
+import qualified Data.String as String
+import Juvix.Library
+
+-- | Keep track of messages during the compilation or REPL.
+-- The implementation is based on
+-- [[https://github.com/UU-ComputerScience/uu-cco/blob/master/uu-cco/src/CCO/Feedback.hs]].
+data Feedback (app :: * -> *) msg a
+  = Success (app msg) a -- Indicate success.
+  | Fail (app msg) -- Indicate a failure.
+
+instance Functor (Feedback app msg) where
+  fmap f (Success msgs x) = Success msgs (f x)
+  fmap _ (Fail msgs) = Fail msgs
+
+instance Monoid (app msg) => Applicative (Feedback app msg) where
+  pure x = Success mempty x
+
+  Success msgs f <*> ax = case ax of
+    Success msgs' x -> Success (msgs <> msgs') (f x)
+    Fail msgs' -> Fail (msgs <> msgs')
+  Fail msgs <*> _ = Fail msgs
+
+instance Monoid (app msg) => Monad (Feedback app msg) where
+  return = pure
+  Success msgs x >>= mf = case mf x of
+    Success msgs' x' -> Success (msgs <> msgs') x'
+    Fail msgs' -> Fail (msgs <> msgs')
+  Fail msgs >>= _ = Fail msgs
+
+instance
+  ( Applicative app,
+    Monoid (app msg),
+    String.IsString msg
+  ) =>
+  Fail.MonadFail (Feedback app msg)
+  where
+  fail = Fail . pure . String.fromString
+
+-- | Monad transformer of Feedback.
+data FeedbackT app msg m a = FeedbackT {runFeedbackT :: m (Feedback app msg a)}
+
+instance Monad m => Functor (FeedbackT app msg m) where
```

Comment:
Can you derive some of these instances automatically (e.g. Functor, Applicative)?